jupyterhub/proxy.py

```diff
@@ -768,10 +769,47 @@ async def api_request(self, path, method='GET', body=None, client=None):
             method=method,
             headers={'Authorization': 'token {}'.format(self.auth_token)},
             body=body,
+            connect_timeout=1,  # default: 20s
+            request_timeout=5,  # default: 20s
         )
-        async with self.semaphore:
-            result = await client.fetch(req)
-            return result
+
+        most_recent_error = None
+        async def _wait_for_api_request():
+            try:
+                async with self.semaphore:
+                    return await client.fetch(req)
+            except HTTPError as e:
+                most_recent_error = e
+                # Retry on potentially transient errors as for example also
+                # suggested for communication witht he k8s api-server
+                #
+                # ref: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#error-codes
+                if e.code in [
+                    429,  # Too many requests
+                    500,  # Internal server error
+                    503,  # Service unavailable
+                    504,  # Gateway timeout
+                    599,  # Network connect timeout error
+                ]:
+                    self.log.warning(
+                        "api_request to the proxy failed with status code {}, retrying...".format(
+                            e.code
+                        )
+                    )
+                    return False  # a falsy return value make exponential_backoff retry
+                else:
+                    self.log.error("api_request to proxy failed: {0}".format(e))
+                    # An unhandled error here will help the hub invoke cleanup logic
+                    raise
+
+        result = await exponential_backoff(
+            _wait_for_api_request,
+            'Repeated api_request to proxy path "{}" failed, most recently with error "{}".'.format(
+                path, most_recent_error
```

Comment:
This string-formatting is evaluated immediately, so the attempt to use most_recent_error here isn't going to work since it will always be None.

I don't think we need it, though, because the most recent error is already logged when it happens, before reaching this failure message.

Reply:
Deal - removed it!